### ISSUE TITLE
Clarification on MMD Loss Calculation for Categorical Distributions

### ISSUE THREAD

**Issue by NafiseHaghshenas** [isjakewong/InfoDiffusion#6]
Hi,

First, I'd like to say thank you for your help and suggestion. It has been really useful.

I have a follow-up question regarding the implementation of Maximum Mean Discrepancy (MMD) loss in the context of categorical distributions.
In the code, I have the following:

a, a_q, mu, log_var = self.encoder(x)

In the case of a categorical distribution, instead of having continuous values like mu (which represents means), we work with the Gumbel-Softmax sampled values and the softmax of the output logits.

Z_gumble_softmax_sample, qy = self.encoder(x)

For the KLD loss in categorical setup, the formula seems to be:

log_ratio = torch.log(qy * categorical_dim + 1e-20)
KLD = torch.sum(qy * log_ratio, dim=-1).mean()

Where qy represents the predicted categorical probabilities.

Now, when calculating the MMD loss, we have:

loss_mmd = compute_mmd(true_samples, mu)

I’m unclear about what true_samples and mu represent in the case of categorical distributions. How should these values be interpreted in this context? Specifically, what should mu represent for categorical distributions, and what exactly are true_samples in this case?

Any insights would be greatly appreciated. Thanks!
Nafise

**Comment by isjakewong** [comment]
Use `Z_gumbel_softmax_sample `instead of `mu` in your case since it acts as the continuous embedding of the categorical latent variable, just like `mu `did in the Gaussian setting. The true samples would be one-hot vectors (or their relaxed versions via Gumbel-Softmax) sampled from the uniform distribution since the prior is typically a uniform categorical distribution over the K categories. E.g. `prior_samples = F.gumbel_softmax(torch.log(prior_probs), tau=temperature, hard=False)`
